File: src/_03_BehavioralPattern/_03_08_State/java/after/Student.java
package _03_BehavioralPattern._03_08_State.java.after;

import java.util.HashSet;
import java.util.Set;

public class Student {
  
  private String name;

  private Set<OnlineCourse> onlineCourses = new HashSet<>();
  
  public Student(String name) {
    this.name = name;
  }

  public boolean isAvailable(OnlineCourse onlineCourse) {
    return this.onlineCourses.contains(onlineCourse);
  }

  public void addPrivate(OnlineCourse onlineCourse) {
    this.onlineCourses.add(onlineCourse);
  }

  @Override
  public String toString() {
    return "\nStudent: { "
      + " name: " + this.name
      + " } ";
  }
  
}

File: src/_03_BehavioralPattern/_03_08_State/Student.js
/**
 * @typedef { import("./OnlineCourse").default } OnlineCourse
 */

export default class Student {
  /** @type { string } */
  _name;

  /** @type { Set<OnlineCourse> } */
  _onlineCourses;
  
  /** @param { string } name */
  constructor(name) {
    this._name = name;
    this._onlineCourses = new Set();
  }

  /** @returns { string } */
  getName() {
    return this._name;
  }

  /** @returns { boolean } */
  isAvailable(onlineCourse) {
    return this._onlineCourses.has(onlineCourse);
  }

  /** @param { OnlineCourse } onlineCourse */
  addPrivate(onlineCourse) {
    this._onlineCourses.add(onlineCourse);
  }
}